### ISSUE TITLE
Count based index rotation can be delayed

### ISSUE THREAD

**Issue by mpfz0r** [Graylog2/graylog2-server#11689]
We configured a count based index rotation (>2000) on an index that is not queried on a regular basis.

While the index is constantly written with new data, it isn't queried.
This leads to a situation where the rotation was delayed by up to 13 hours.

The problem comes from the fact that the index stats API is giving us stale results:
```
2021-11-24 16:03:37,145 DEBUG: org.graylog2.indexer.rotation.strategies.AbstractRotationStrategy - Rotation strategy result: Number of messages in <graylog_129> (117) is lower than the limit (2,000). Not doing anything. 
2021-11-24 16:03:47,150 DEBUG: org.graylog2.indexer.rotation.strategies.AbstractRotationStrategy - Rotation strategy result: Number of messages in <graylog_129> (117) is lower than the limit (2,000). Not doing anything. 
2021-11-24 16:03:57,144 DEBUG: org.graylog2.indexer.rotation.strategies.AbstractRotationStrategy - Rotation strategy result: Number of messages in <graylog_129> (117) is lower than the limit (2,000). Not doing anything. 
2021-11-24 16:04:07,145 DEBUG: org.graylog2.indexer.rotation.strategies.AbstractRotationStrategy - Rotation strategy result: Number of messages in <graylog_129> (117) is lower than the limit (2,000). Not doing anything. 
```
Only after manually issuing a flush:
`curl  -XPOST  http://localhost:9200/graylog_129/_flush`

It will see the actual document count and start rotating:

```
2021-11-24 16:04:47,156 DEBUG: org.graylog2.indexer.rotation.strategies.AbstractRotationStrategy - Rotation strategy result: Number of messages in <graylog_129> (2,975) is higher than the limit (2,000). Pointing deflector to new index now! 
```

This probably comes from this change in 7.0:

https://www.elastic.co/guide/en/elasticsearch/reference/7.17/breaking-changes-7.0.html#_skipped_background_refresh_on_search_idle_shards
```
Shards belonging to an index that does not have an explicit index.refresh_interval configured
will no longer refresh in the background once the shard becomes "search idle", 
ie the shard hasn’t seen any search traffic for index.search.idle.after seconds (defaults to 30s). 
Searches that access a search idle shard will be "parked" until the next refresh happens.
Indexing requests with wait_for_refresh will also trigger a background refresh.
```

## Solutions
 - Configure `index.refresh_interval` or `index.search.idle.after`
 - Call `_flush` before running a `_stats`query   (this might be too expensive, because the index rotation runs every 10s)

## Versions
* Elasticsearch Version: 7.10
* OpenSearch 2.4.1

**Comment by bernd** [comment]
@mpfz0r Some users override the default refresh interval for index performance reasons via custom index templates. That means we cannot use a hardcoded value here because that would override the user settings.

**Comment by mpfz0r** [comment]
@bernd 
hmm, but which setting would be effective then?

**Comment by bernd** [comment]
> @bernd hmm, but which setting would be effective then?

@mpfz0r 
The ones from your diff. Customers have custom index templates. If we don't set `refresh_interval` in the index settings, the setting will be taken from the template. But since you set this explicitly for the index, the template values will not be used. At least that's how I understand it. :slightly_smiling_face:

**Comment by mpfz0r** [comment]
@bernd so could we create a default template then, which the customer can overwrite?
or do you have any other suggestions?

**Comment by bernd** [comment]
> @bernd so could we create a default template then, which the customer can overwrite? or do you have any other suggestions?

@mpfz0r 
Aye, we would have to add the refresh setting to our default template(s). :+1: When users install a custom template with higher precedence, their values will be used. Since this setting is also applied to write indices, we have to check the default value that OpenSearch.

Could we alternatively apply the setting to an index only once we rotate it and it doesn't receive writes anymore?

**Comment by boosty** [comment]
> Could we alternatively apply the setting to an index only once we rotate it and it doesn't receive writes anymore?

I thought the `refresh_interval` is only interesting during indexin. I think it should be a no-op on read-only indices, but maybe someone with more ES/OpenSearch knowledge could confirm.

**Comment by mpfz0r** [comment]
> > Could we alternatively apply the setting to an index only once we rotate it and it doesn't receive writes anymore?
> 
> I thought the `refresh_interval` is only interesting during indexing. 

Correct, applying it after rotation is pointless. The reason for the issue being that we can't rotate, because we get wrong numbers.

**Comment by bernd** [comment]
> > > Could we alternatively apply the setting to an index only once we rotate it and it doesn't receive writes anymore?
> > 
> > 
> > I thought the `refresh_interval` is only interesting during indexing.
> 
> Correct, applying it after rotation is pointless. The reason for the issue being that we can't rotate, because we get wrong numbers.

Ah, sorry. I got that part wrong.

**Comment by mpfz0r** [comment]
This does the trick
```
diff --git graylog2-server/src/main/java/org/graylog2/indexer/IndexMapping.java graylog2-server/src/main/java/org/graylog2/indexer/IndexMapping.java
index 8913de9c03..5e05f38112 100644
--- graylog2-server/src/main/java/org/graylog2/indexer/IndexMapping.java
+++ graylog2-server/src/main/java/org/graylog2/indexer/IndexMapping.java
@@ -21,7 +21,6 @@ import com.google.common.collect.ImmutableMap;
 import org.graylog2.indexer.indexset.IndexSetConfig;
 import org.graylog2.plugin.Message;
 
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
@@ -31,6 +30,7 @@ import java.util.Map;
  */
 public abstract class IndexMapping implements IndexMappingTemplate {
     public static final String TYPE_MESSAGE = "message";
+    final String indexRefreshInterval = "30s"; // TODO: Index refresh interval must be configurable
 
     @Override
     public Map<String, Object> toTemplate(IndexSetConfig indexSetConfig, String indexPattern, int order) {
@@ -44,9 +44,10 @@ public abstract class IndexMapping implements IndexMappingTemplate {
     }
 
     public Map<String, Object> messageTemplate(final String template, final String analyzer, final int order) {
-        final Map<String, Object> settings = Collections.singletonMap(
-                "analysis", Collections.singletonMap("analyzer", analyzerKeyword())
-                );
+        final Map<String, Object> settings = Map.of(
+                "analysis", Map.of("analyzer", analyzerKeyword()),
+                "index.refresh_interval", indexRefreshInterval
+        );
         final Map<String, Object> mappings = mapping(analyzer);
 
         return createTemplate(template, order, settings, mappings);
```
